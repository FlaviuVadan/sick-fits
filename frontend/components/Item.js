// use rccp for basic React component template

import Link from 'next/link';
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import DeleteItem from './DeleteItem';
import AddToCart from './AddToCart';
import formatMoney from '../lib/formatMoney';


class Item extends Component {
	static propTypes = {
		// notice, can do the following for building up a custom object:
		// items: PropTypes.shape({
		// 	title: PropTypes.string.isRequired,
		// 	price: PropTypes.number.isRequired,
		items: PropTypes.object.isRequired
	};

	render() {
		// destructure the item, specifically
		const { item } = this.props;
		return (
			<ItemStyles>
				{/*a trick to check if the image is present on the item and source it if it's there*/}
				{item.image && <img src={item.image} alt={item.title}/>}

				<Title>
					<Link href={{
						pathname: '/item',
						query: {
							id: item.id
						},
					}}><a>{item.title}</a></Link>
				</Title>

				<PriceTag>{formatMoney(item.price)}</PriceTag>

				<p>{item.description}</p>

				<div className="buttonList">
					<Link href={{
						pathname: '/update',
						query: {
							id: item.id,
						}
					}}><a>Edit</a></Link>
					<AddToCart id={item.id}/>
					<DeleteItem id={item.id}>Delete</DeleteItem>
				</div>

			</ItemStyles>
		);
	}
}

Item.propTypes = {};

export default Item;