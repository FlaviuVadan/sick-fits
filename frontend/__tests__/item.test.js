// shallow render top-level component, do not go to children components because they have individual tests
import { shallow } from 'enzyme';
import Item from '../components/Item';

const fakeItem = {
	id: 'ABC123',
	title: 'An item',
	price: '5000',
	description: 'This is an item',
	image: 'item.jpg',
	largeImage: 'itemLarge.jpg',
};

// can also separate the it()s in describe() into multiple tests for checking different component parts
// e.g PriceTag, Title a, etc.
describe('<Item/>', () => {
	it('renders and displays properly', () => {
		const wrapper = shallow(<Item item={fakeItem}/>);
		// to display the full wrapper HTML component, use wrapper.debug()
		// console.log(wrapper.debug());

		const priceTag = wrapper.find('PriceTag');
		// can perform a dive on shallow-rendered components
		// console.log(priceTag.dive().text());
		expect(priceTag.children().text()).toBe('$50');

		// can get Title along with anchor tag <a/>
		const title = wrapper.find('Title a');
		expect(title.text()).toBe(fakeItem.title);

		const image = wrapper.find('img');
		expect(image.props().src).toEqual(fakeItem.image);
		expect(image.props().alt).toEqual(fakeItem.title);
	});

	it('renders out buttons properly', () => {
		const wrapper = shallow(<Item item={fakeItem}/>);
		const buttonList = wrapper.find('.buttonList');
		expect(buttonList.children()).toHaveLength(3);
		expect(buttonList.find('Link').exists()).toBeTruthy();
		expect(buttonList.find('AddToCart').exists()).toBeTruthy();
		expect(buttonList.find('DeleteItem').exists()).toBeTruthy();
	});
});