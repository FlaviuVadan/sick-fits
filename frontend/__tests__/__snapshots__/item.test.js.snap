// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item/> renders and matches the snapshot 1`] = `
<ItemStyles__Item>
  <img
    alt="An item"
    src="item.jpg"
  />
  <Title>
    <Link
      href={
        Object {
          "pathname": "/item",
          "query": Object {
            "id": "ABC123",
          },
        }
      }
    >
      <a>
        An item
      </a>
    </Link>
  </Title>
  <PriceTag>
    $50
  </PriceTag>
  <p>
    This is an item
  </p>
  <div
    className="buttonList"
  >
    <Link
      href={
        Object {
          "pathname": "/update",
          "query": Object {
            "id": "ABC123",
          },
        }
      }
    >
      <a>
        Edit
      </a>
    </Link>
    <AddToCart
      id="ABC123"
    />
    <DeleteItem
      id="ABC123"
    >
      Delete
    </DeleteItem>
  </div>
</ItemStyles__Item>
`;
